### ISSUE TITLE
Using Homebrew installed resources in Moonshine

### ISSUE THREAD

**Issue by piotrzarzycki21** [prominic/Moonshine-IDE#493]
@carlosrovira decided to give a try to Moonshine. He started [thread](http://apache-royale-development.20373.n8.nabble.com/Trying-Moonshine-2-0-td9316.html) on Royale dev list. 

His first approach after launching 2.0.0 was using installed resources by [Homebrew](https://brew.sh/). Can someone with more knowledge on Mac platform can look into that. All those resources installed by Homebrew are hidden - can we make a test and make them visible, point in the Moonshine for example ANT and see whether we can use it?

If it will work - Carlos suggested pretty interested solution in the thread on dev list - maybe we should implement that ?

**Comment by JoelProminic** [comment]
I did a test with Ant installed from Brew.

    brew install ant

To browse to a hidden path in a macOS file chooser, use CMD-Shift-G.  This is a built-in macOS shortcut that also works in Finder.  I like the idea of having a button for this instead - we can check to see if this is something we can easily add.

Note:  I needed to set the Ant Home path to "/usr/local/Cellar/ant/1.10.5/libexec" in order to work around this issue:  https://stackoverflow.com/questions/33603454/apache-ant-could-not-find-or-load-main-class-org-apache-tools-ant-launch-launche

Unfortunately, on recent versions of macOS, you will see an error like this:

    /usr/local/Cellar/ant/1.10.5/libexec/bin/ant: /bin/sh: bad interpreter: Operation not permitted

This triggers when bash scripts are executed by an app running in the Apple Sandbox.   The workaround we are using for this is to download Ant to ~/Downloads, which has different restrictions.

**Comment by piotrzarzycki21** [comment]
> This triggers when bash scripts are executed by an app running in the Apple Sandbox. The workaround we are using for this is to download Ant to ~/Downloads, which has different restrictions.

In the other words because we are working in a SandBox environment there is no way using Homebrew stuff. Use will have to use non restricted stuff placed to Downloads. Do we have some more workarounds for that problem ?

**Comment by carlosrovira** [comment]
Hi,

I could  successfully set ant with @JoelProminic trick (CMD+Shift+G), but as you said would be great to have a more visible button. Then I tried to run a build.xml script and I get the before mentioned "bad interpreter: Operation not permitted" message. 

A quick search showed me this: https://blog.omgmog.net/post/a-possible-solution-to-bad-interpreter-operation-not-permitted-error-on-mac-os-x-lion/
Don't know if this is a real solution or not, maybe it deserves a bit of more investigation, hope you could get to some solution for this problem
thanks!

**Comment by rat-moonshine** [comment]
Hei @carlosrovira , thank you for your feedback. 

On the operation-not-permitted subject/link,

> A quick search showed me this: https://blog.omgmog.net/post/a-possible-solution-to-bad-interpreter-operation-not-permitted-error-on-mac-os-x-lion/
> Don't know if this is a real solution or not, maybe it deserves a bit of more investigation, hope you could get to some solution for this problem

Unfortunately, it is not permitted to execute `xattr` command in sandbox application to remove any file's extended attributes (i.e. `com.apple.quarantine`), unless the application using special entitlement (Apple never let us use those special entitlements, however).

Furthermore, as the above link mentioned if you download something through browser, OSX adds an extended attribute to it, if you ran such executable from ~/Downloads folder, you may hit by same 'operation-not-permitted' error; unless you remove them manually. Good news is, if you download the files through an application, like Apache SDK Installer or Moonshine SDK Installer, you will not get this extended attribute added.

**Comment by carlosrovira** [comment]
Hi, 

I'm thinking that VSCode can run ANT, MAVEN and other tools from HomeBrew, maybe is just a matter of ask them or see in his code how they do. I think this kind of things are important since I think users should not depend on download through Moonshine, the normal scenario will be users will have some (or all) needed tools. Thanks!

**Comment by piotrzarzycki21** [comment]
@rat-moonshine  Correct me if I'm wrong but in general we have all sort of problems, cause we are providing Moonshine trough app store. VSCode can be downloaded directly - Am I correct ?

**Comment by rat-moonshine** [comment]
Well, yes, you're right @piotrzarzycki21 . Due to Apple sandbox, we requires to manage many restrictions in available alternate ways when other IDEs do not fall into these restrictions, simply because they do not run in sandbox.

**Comment by carlosrovira** [comment]
I suppose Apple should have some solutions to solve this problems, if not, I think Apple Store gives more problems than it solves and maybe a you should provide versions on you own way. Maybe this is one of the reasons we don't see other packages distributed via Apple Store.

**Comment by piotrzarzycki21** [comment]
It looks like w rather won't be able to do anything with that stuff. I will close it soon unless someone have some more thoughts.

**Comment by carlosrovira** [comment]
Hi Piotr, it's a pity since it make for me almost not possible to use. But I understand that each platform can decide what's go in or not. Thanks.

**Comment by piotrzarzycki21** [comment]
Blame AppStore :) Yes I'm also not satisfied about that. Maybe some day we will go away from AppStore, but it is not going to happen too soon.

**Comment by piotrzarzycki21** [comment]
Anyone who do not have problems with not using Homebrew installed stuff but install software using Moonshine can use IDE without the problem. You are using Intellij with their Java provided with IDE, so this is similar situation. We would probably included java as part of the IDE, but we simply cannot (because of the AppStore), so user have to install it.

**Comment by carlosrovira** [comment]
I'm not using IntelliJ nor VSCode with bundled Java. We have installed on our own since we need to have control of our environment. IDEs are just a tool that are on top of that to ease development, but can impose a scenario. Anyway, is legit that you choose one way or another. Let me know when you have a version outside App Store that has this solved. Thanks :)

**Comment by piotrzarzycki21** [comment]
> I'm not using IntelliJ nor VSCode with bundled Java. We have installed on our own since we need to have control of our environment. IDEs are just a tool that are on top of that to ease development, but can impose a scenario. Anyway, is legit that you choose one way or another. Let me know when you have a version outside App Store that has this solved. Thanks :)

@carlosrovira If you have ever use intellij it coming with bundled Java inside. It doesn't matter what kind of Java version you have installed - Intellij using it's internal. The same is with Ant, so Moonshine approach is similar one. The difference is that we are providing clear view what you are going to install with intellij you didn't have too much option to change that. Example of Java path which has been run by Intellij:
![java_intellij](https://user-images.githubusercontent.com/24554795/55570584-006add80-5704-11e9-8a80-4b17f27c5e9b.png)

**Comment by carlosrovira** [comment]
Hi Piotr, you can avoid using IntelliJ provided Java, I'm referring to this, if you want to use your own installed version of Java in IntelliJ, in our case home-brew installation. Mac instructions provided by JetBrains in its own page:

> Our latest IDE versions come with the bundled custom JDK 8 which contains the fixes for most known OpenJDK and Oracle JDK bugs. Should you need to use a different Java version, please refer to this blog post.
> If you override IDE JDK version, its path is stored in <product>.jdk file located in the config folder (idea.jdk for IntelliJ IDEA, pycharm.jdk for PyCharm, etc). Delete this file or change the path inside the file manually in case IDE no longer starts and you can't change it via the menu.
> If IDE doesn't start and this file doesn't exist, create it manually and specify Java path to use (Java home location), for example:
> /Library/Java/JavaVirtualMachines/jdk1.8.0_112.jdk

So IntelliJ does not force you to run under the Java 8 jdk they provide internaly if you don't want, since they know an IDE must be flexible in this kind of things to be able to use by anyone out there. The problem is we all have very different situations.

Maybe unrelated (or not), but, a part from that (use our own java 8 version), we use to switch from Java 6,7 and 8 per project, and use "jenv", a very convenient tool to define easily a default JDK per system folder, so when we go to a concrete project Mac knows in that folder the java version to use in a the jenv one defined. jenv is installed by homebrew too.

**Comment by piotrzarzycki21** [comment]
> So IntelliJ does not force you to run under the Java 8 jdk they provide internally if you don't want, since they know an IDE must be flexible in this kind of things to be able to use by anyone out there. The problem is we all have very different situations.

We also do allow for that, unfortunately AppStore limits us to use java from specific location and this location is not in a Homebrew installation folder.

**Comment by piotrzarzycki21** [comment]
Moonshine 2.2.0 has been released with non-sandbox version where you should be able to use Homebrew stuff.

**Comment by carlosrovira** [comment]
Hi Piotr and Justin,
just installed new version, and go to ANT home config, click config, then change and set up with

/usr/local/Cellar/ant/1.10.5/libexec/bin

as well tried :

/usr/local/Cellar/ant/1.10.5/bin/

(since in both places we have ant executable

then save

the result is that Apache ANT didn't change to green tick. Is that right? other things like Git or Subversion are showing a green tick

thanks

**Comment by JoelProminic** [comment]
Hi @carlosrovira,

The Moonshine setting is supposed to represent the path to use for ANT_HOME.  The binary should be at $ANT_HOME/bin/ant.  So, the correct path in this case is:

    /usr/local/Cellar/ant/1.10.5/libexec

NOTE:  /usr/local/Cellar/ant/1.10.5/ did not work for me.  I got an error like the one below.  If you get the same error, try going through the fixes here:  https://stackoverflow.com/questions/33603454/apache-ant-could-not-find-or-load-main-class-org-apache-tools-ant-launch-launche

    Error: Could not find or load main class org.apache.tools.ant.launch.Launcher

The Settings interface should make the expected target more clear, and we'll want to revisit this for the other SDKs as well.

Also, I see that the Getting Started page doesn't update when I set the path correctly.  I'll create a separate issue for this.  Ant should start working once once you have set the path, though.
